File: bot/tests/sqlite_test.py
import os
import unittest
from datetime import datetime
from typing import Any, Dict, List, Optional
from db.sqlite import BotDB


class TestBotDB(unittest.TestCase):
    def setUp(self):
        self.db_path = "test_bot.db"
        if os.path.exists(self.db_path):
            os.remove(self.db_path)
        self.db = BotDB(db_path=self.db_path)

    def tearDown(self):
        if os.path.exists(self.db_path):
            os.remove(self.db_path)

    def test_trusted_users(self):
        user_id = 123
        admin_id = 456
        self.db.trust_user(user_id, admin_id)
        self.assertTrue(self.db.is_user_trusted(user_id))

        user: Optional[Dict[str, Any]] = self.db.get_trusted_user(user_id)
        self.assertIsNotNone(user)
        if user:
            self.assertEqual(user["user_id"], user_id)
            self.assertEqual(user["by"], admin_id)

        self.db.untrust_user(user_id)
        self.assertFalse(self.db.is_user_trusted(user_id))

    def test_buktopuha(self):
        user_id = 1
        user_meta = {"id": 1, "first_name": "Test"}
        self.db.add_buktopuha_player(user_id, user_meta, score=10)

        player: Optional[Dict[str, Any]] = self.db.find_buktopuha_player(user_id)
        self.assertIsNotNone(player)
        if player:
            self.assertEqual(player["_id"], user_id)
            self.assertEqual(player["total_score"], 10)
            self.assertEqual(player["win_counter"], 1)
            self.assertEqual(player["game_counter"], 0)

        self.db.inc_buktopuha_game_counter(user_id)
        player = self.db.find_buktopuha_player(user_id)
        self.assertIsNotNone(player)
        if player:
            self.assertEqual(player["game_counter"], 1)

        self.db.inc_buktopuha_win(user_id, score=5)
        player = self.db.find_buktopuha_player(user_id)
        self.assertIsNotNone(player)
        if player:
            self.assertEqual(player["win_counter"], 2)
            self.assertEqual(player["total_score"], 15)

        players: List[Dict[str, Any]] = self.db.get_all_buktopuha_players()
        self.assertEqual(len(players), 1)

        self.db.remove_buktopuha_player(user_id)
        self.assertIsNone(self.db.find_buktopuha_player(user_id))

    def test_towel_quarantine(self):
        user_id = 789
        self.db.add_quarantine_user(user_id, quarantine_time_min=60)

        user: Optional[Dict[str, Any]] = self.db.find_quarantine_user(user_id)
        self.assertIsNotNone(user)
        if user:
            self.assertEqual(user["_id"], user_id)
            self.assertEqual(user["rel_messages"], [])

        self.db.add_quarantine_rel_message(user_id, 1001)
        user = self.db.find_quarantine_user(user_id)
        self.assertIsNotNone(user)
        if user:
            self.assertIn(1001, user["rel_messages"])

        users: List[Dict[str, Any]] = self.db.find_all_quarantine_users()
        self.assertEqual(len(users), 1)

        self.db.delete_quarantine_user(user_id)
        self.assertIsNone(self.db.find_quarantine_user(user_id))

    def test_since_topics(self):
        topic = "Rust"
        now = datetime.now()
        self.db.update_since_topic(topic, now, 1)

        t: Dict[str, Any] = self.db.get_since_topic(topic)
        self.assertEqual(t["topic"], topic.lower())
        self.assertEqual(t["count"], 1)

        self.db.update_since_topic(topic, now, 1)
        t = self.db.get_since_topic(topic)
        self.assertEqual(t["count"], 2)

        topics: List[Dict[str, Any]] = self.db.get_all_since_topics(limit=10)
        self.assertEqual(len(topics), 1)

    def test_roll_hussars(self):
        user_id = 2
        user_meta = {"id": 2, "username": "hussar"}
        self.db.add_hussar(user_id, user_meta)

        h: Optional[Dict[str, Any]] = self.db.find_hussar(user_id)
        self.assertIsNotNone(h)
        if h:
            self.assertEqual(h["_id"], user_id)
            self.assertEqual(h["shot_counter"], 0)

        self.db.hussar_miss(user_id)
        h = self.db.find_hussar(user_id)
        self.assertIsNotNone(h)
        if h:
            self.assertEqual(h["shot_counter"], 1)
            self.assertEqual(h["miss_counter"], 1)

        self.db.hussar_dead(user_id, mute_min=60)
        h = self.db.find_hussar(user_id)
        self.assertIsNotNone(h)
        if h:
            self.assertEqual(h["dead_counter"], 1)
            self.assertEqual(h["total_time_in_club"], 3600)

        hussars: List[Dict[str, Any]] = self.db.get_all_hussars()
        self.assertEqual(len(hussars), 1)

    def test_prism_words(self):
        word = "hello"
        self.db.add_prism_word(word)
        self.db.add_prism_word(word)

        words: List[Dict[str, Any]] = self.db.get_all_prism_words()
        self.assertEqual(len(words), 1)
        self.assertEqual(words[0]["word"], word)
        self.assertEqual(words[0]["count"], 2)

    def test_peninsula_users(self):
        user_id = 3
        user_meta = {"id": 3, "name": "Peninsula"}
        self.db.add_peninsula_user(user_id, user_meta)

        users: List[Dict[str, Any]] = self.db.get_best_peninsulas(n=10)
        self.assertEqual(len(users), 1)
        self.assertEqual(users[0]["_id"], user_id)
        self.assertEqual(users[0]["meta"], user_meta)

    def test_aoc_data(self):
        data = {"members": {"1": {"name": "AOC User"}}}
        self.db.update_aoc_data(data)

        stored: Optional[Dict[str, Any]] = self.db.get_aoc_data()
        self.assertEqual(stored, data)

        self.db.remove_all_aoc_data()
        self.assertIsNone(self.db.get_aoc_data())


if __name__ == "__main__":
    unittest.main()

File: internal/db/sqlite_test.go
package db

import (
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func newTestDB(t *testing.T) *DB {
	t.Helper()
	f, err := os.CreateTemp("", "test_bot_*.db")
	require.NoError(t, err)
	path := f.Name()
	_ = f.Close()
	t.Cleanup(func() { os.Remove(path) })

	d, err := New(path)
	require.NoError(t, err)
	t.Cleanup(func() { _ = d.Close() })
	return d
}

func TestTrustedUsers(t *testing.T) {
	d := newTestDB(t)

	trusted, err := d.IsUserTrusted(123)
	require.NoError(t, err)
	assert.False(t, trusted)

	require.NoError(t, d.TrustUser(123, 999))
	trusted, err = d.IsUserTrusted(123)
	require.NoError(t, err)
	assert.True(t, trusted)

	require.NoError(t, d.UntrustUser(123))
	trusted, err = d.IsUserTrusted(123)
	require.NoError(t, err)
	assert.False(t, trusted)
}

func TestQuarantine(t *testing.T) {
	d := newTestDB(t)

	// Add quarantine user
	require.NoError(t, d.AddQuarantineUser(42, 60))

	u, err := d.FindQuarantineUser(42)
	require.NoError(t, err)
	require.NotNil(t, u)
	assert.Equal(t, int64(42), u.UserID)
	assert.False(t, u.IsExpired())

	// Add related message
	require.NoError(t, d.AddQuarantineRelMessage(42, 100))
	require.NoError(t, d.AddQuarantineRelMessage(42, 101))
	// Dedup
	require.NoError(t, d.AddQuarantineRelMessage(42, 100))

	u, err = d.FindQuarantineUser(42)
	require.NoError(t, err)
	assert.Equal(t, []int{100, 101}, u.MessageIDs())

	// Delete
	require.NoError(t, d.DeleteQuarantineUser(42))
	u, err = d.FindQuarantineUser(42)
	assert.Error(t, err)
	assert.Nil(t, u)

	// Delete all
	require.NoError(t, d.AddQuarantineUser(1, 10))
	require.NoError(t, d.AddQuarantineUser(2, 10))
	require.NoError(t, d.DeleteAllQuarantineUsers())
	users, err := d.FindAllQuarantineUsers()
	require.NoError(t, err)
	assert.Empty(t, users)
}

func TestQuarantineIdempotent(t *testing.T) {
	d := newTestDB(t)

	require.NoError(t, d.AddQuarantineUser(42, 60))
	// Adding again should be no-op
	require.NoError(t, d.AddQuarantineUser(42, 120))

	u, _ := d.FindQuarantineUser(42)
	require.NotNil(t, u)
}

func TestHussars(t *testing.T) {
	d := newTestDB(t)

	// Initially empty
	hussars, err := d.GetAllHussars()
	require.NoError(t, err)
	assert.Empty(t, hussars)

	// Add hussar
	meta := `{"username":"testuser","first_name":"Test"}`
	require.NoError(t, d.AddHussar(123, meta))

	h, err := d.FindHussar(123)
	require.NoError(t, err)
	require.NotNil(t, h)
	assert.Equal(t, "testuser", h.Username())
	assert.Equal(t, 0, h.ShotCounter)

	// Miss
	require.NoError(t, d.HussarMiss(123))
	h, _ = d.FindHussar(123)
	assert.Equal(t, 1, h.ShotCounter)
	assert.Equal(t, 1, h.MissCounter)
	assert.Equal(t, 0, h.DeadCounter)

	// Dead
	require.NoError(t, d.HussarDead(123, 960))
	h, _ = d.FindHussar(123)
	assert.Equal(t, 2, h.ShotCounter)
	assert.Equal(t, 1, h.DeadCounter)
	assert.Equal(t, 960*60, h.TotalTimeInClub)

	// Username fallback
	meta2 := `{"first_name":"John","last_name":"Doe"}`
	require.NoError(t, d.AddHussar(456, meta2))
	h2, _ := d.FindHussar(456)
	assert.Equal(t, "John Doe", h2.Username())

	// Remove
	require.NoError(t, d.RemoveHussar(123))
	_, err = d.FindHussar(123)
	assert.Error(t, err)

	// Wipe
	require.NoError(t, d.RemoveAllHussars())
	hussars, _ = d.GetAllHussars()
	assert.Empty(t, hussars)
}

func TestSinceTopics(t *testing.T) {
	d := newTestDB(t)

	// First mention creates
	require.NoError(t, d.UpsertSinceTopic("golang"))
	topic, err := d.GetSinceTopic("golang")
	require.NoError(t, err)
	assert.Equal(t, 1, topic.Count)

	// Second mention increments
	require.NoError(t, d.UpsertSinceTopic("golang"))
	topic, _ = d.GetSinceTopic("golang")
	assert.Equal(t, 2, topic.Count)

	// List
	require.NoError(t, d.UpsertSinceTopic("rust"))
	topics, err := d.GetAllSinceTopics(10)
	require.NoError(t, err)
	assert.Len(t, topics, 2)
	assert.Equal(t, "golang", topics[0].Topic) // golang has 2, rust has 1
}

func TestPrismWords(t *testing.T) {
	d := newTestDB(t)

	require.NoError(t, d.AddPrismWord("hello"))
	require.NoError(t, d.AddPrismWord("hello"))
	require.NoError(t, d.AddPrismWord("world"))

	words, err := d.GetAllPrismWords()
	require.NoError(t, err)
	assert.Len(t, words, 2)
	assert.Equal(t, "hello", words[0].Word)
	assert.Equal(t, 2, words[0].Count)
}
